### DISCUSSION TITLE
Uploading from a saved hexfile via megaTinyCore without having source code.

### DISCUSSION THREAD

**Discussion by MitchBogart** [SpenceKonde/megaTinyCore#670] [Q&A]
I need to give an entity just the hex file to upload (via UPDI).  Program uploads quickly (UPDI to an ATTiny402).
Is there a batch file than can accompany the saved hex file for a 3rd party to load and test? 
Are there any command line call(s) to megaTinyCore or Menu selections to accomplish that?
I can share the exact log from "Upload Using Programmer".  It always compiles first, thus needing the source code.
I would make and include a working CH340 + Schottky diode assembly.
Thank you!

**Comment by SpenceKonde** [comment]
the output shown with "show verbose output during: Uploads" will give you the command needed to invoke SerialUPDI 

For my system, it might look like: 

C:\Users\Spence\Documents\Arduino\hardware\megaTinyCore\megaavr/tools/python3/python3 -u C:\Users\Spence\Documents\Arduino\hardware\megaTinyCore\megaavr/tools/prog.py -t uart -u COM19 -b 230400 -d attiny412 --fuses 2:0x02 6:0x04 8:0x00 -fC:\Users\Spence\AppData\Local\Temp\arduino_build_54126/trashsketch.ino.hex -a write -v 

If you installed it by board manager you likely see something very different for your python path.... (the platform.txt where the recipe to construct the command lies actually gets edited by a script when the core is released, so released versions look for it to have been installed by board manager, but manual installations expect a python3 folder inside of tools.

Likely the more problematic part is that the entity would require SerialUPDI and it would need to be able to run. We provide our own copies of all libraries it uses, so it should require only python and the tools folder (with all the magic contained in libs including all libraries required). That sort of standalone solution is likely the most reliable, but it depends on being able to get a bulky set of tools instead of just serialUPDI to this entity. Depeding on the nature of the entity and the security regime in place that may be problematic. Technical competence of the entity, the context of entity's computing platform have an impact.

If the entity is being supplied with physical media and runs a windows system, a thumbdrive with the contents of a manual installation tool directory would virtually guarantee success - provided security measures didn't prevent execution of binaries on thumbdrives (as yo may know, one tactic used in espionage, industrial and otherwise, is scattering malware laced thumbdrives around a target. There are places that have gone so far as removing all USB ports they can, filling them in with glue, that kind of crazy shit - some institutions are really serious about this. Medical places in particular, deending on how data breach fines are calculated. In some cases, the fine is per piece of information (ie, the learn the doctor's office sees someone named Spence, that's one piece. If they know the last name too, that's two pieces. When tha change occurred, medical customers of software suddenly acquired a newfound concern for security, and started taking dramatic measures like that).

(yes, apparently the first sketch window I grabbed was "trashsketch.ino" - I've bbeen trying to clearly deliniate which sketches can be safely deleted because the sketchbook folder had gotten overwhelmed with sketches named only by date - with no indicatiom of  whether they blink saved as so that I could do export compiled binary, or a complicated sketch that had hours of valuble work in it.

**Comment by MitchBogart** [comment]
Perfect, Spence!  I set verbose upload and was initially concerned as I was
scrolling backward through the whole log.
Then when I reached the beginning I realized you had thoughtfully placed
the exact upload action in a single command line first and the rest was
"just" helpful translation!
It is exactly what I was looking for.
I am dealing with a very experienced entity.  Therefore,
I will provide them with this log with the highlighted command line, the
hex file, the python folder and the tools folder.
I'll also describe and quote the CH340-based SerialUPDI device and your
description of the conveniently sized specific Schottky diode mod.
Thank you!
Best regards,
Mitch

This is the command line (I bracketed it in the log with the 2 bold lines):
Sketch uses 1220 bytes (29%) of program storage space. Maximum is 4096
bytes.
Global variables use 18 bytes (7%) of dynamic memory, leaving 238 bytes for
local variables. Maximum is 256 bytes.

***<Mitch>Next line is the beginning of the upload command line ***
C:\Users\Mitch
Bogart\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1/python3
-u C:\Users\Mitch
Bogart\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.5.10/tools/prog.py
-t uart -u COM5 -b 460800 -d attiny402 --fuses 2:0x02 6:0x04 8:0x00
-fC:\Users\MITCHB~1\AppData\Local\Temp\arduino_build_750162/Ctrlr2_2_3wire_wLight_ATtiny402_ex3_wStaggerLoad-option.ino.hex
-a write -v
***<Mitch>Previous line was the end of the upload command line ***

SerialUPDI
UPDI programming for Arduino using a serial adapter
Based on pymcuprog, with significant modifications
By Quentin Bolsee and Spence Konde
Version 1.2.3 - Jan 2022
Using serial port COM5 at 460800 baud.
Target: attiny402
Set fuses: ['2:0x02', '6:0x04', '8:0x00']
Action: write
File:
C:\Users\MITCHB~1\AppData\Local\Temp\arduino_build_750162/Ctrlr2_2_3wire_wLight_ATtiny402_ex3_wStaggerLoad-option.ino.hex
......

On Mon, Mar 21, 2022 at 11:47 PM Spence Konde (aka Dr. Azzy) <
***@***.***> wrote:

> the output shown with "show verbose output during: Uploads" will give you
> the command needed to invoke SerialUPDI
>
> For my system, it might look like:
>
> C:\Users\Spence\Documents\Arduino\hardware\megaTinyCore\megaavr/tools/python3/python3
> -u
> C:\Users\Spence\Documents\Arduino\hardware\megaTinyCore\megaavr/tools/prog.py
> -t uart -u COM19 -b 230400 -d attiny412 --fuses 2:0x02 6:0x04 8:0x00
> -fC:\Users\Spence\AppData\Local\Temp\arduino_build_54126/trashsketch.ino.hex
> -a write -v
>
> If you installed it by board manager you likely see something very
> different for your python path.... (the platform.txt where the recipe to
> construct the command lies actually gets edited by a script when the core
> is released, so released versions look for it to have been installed by
> board manager, but manual installations expect a python3 folder inside of
> tools.
>
> Likely the more problematic part is that the entity would require
> SerialUPDI and it would need to be able to run. We provide our own copies
> of all libraries it uses, so it should require only python and the tools
> folder (with all the magic contained in libs including all libraries
> required). That sort of standalone solution is likely the most reliable,
> but it depends on being able to get a bulky set of tools instead of just
> serialUPDI to this entity. Depeding on the nature of the entity and the
> security regime in place that may be problematic. Technical competence of
> the entity, the context of entity's computing platform have an impact.
>
> If the entity is being supplied with physical media and runs a windows
> system, a thumbdrive with the contents of a manual installation tool
> directory would virtually guarantee success - provided security measures
> didn't prevent execution of binaries on thumbdrives (as yo may know, one
> tactic used in espionage, industrial and otherwise, is scattering malware
> laced thumbdrives around a target. There are places that have gone so far
> as removing all USB ports they can, filling them in with glue, that kind of
> crazy shit - some institutions are really serious about this. Medical
> places in particular, deending on how data breach fines are calculated. In
> some cases, the fine is per piece of information (ie, the learn the
> doctor's office sees someone named Spence, that's one piece. If they know
> the last name too, that's two pieces. When tha change occurred, medical
> customers of software suddenly acquired a newfound concern for security,
> and started taking dramatic measures like that).
>
> (yes, apparently the first sketch window I grabbed was "trashsketch.ino" -
> I've bbeen trying to clearly deliniate which sketches can be safely deleted
> because the sketchbook folder had gotten overwhelmed with sketches named
> only by date - with no indicatiom of whether they blink saved as so that I
> could do export compiled binary, or a complicated sketch that had hours of
> valuble work in it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/670#discussioncomment-2411689>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYEGUBTEFMAE2GC6Z6R5H5LVBE7EXANCNFSM5RHZ35OQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

Sketch uses 1220 bytes (29%) of program storage space. Maximum is 4096 bytes.
Global variables use 18 bytes (7%) of dynamic memory, leaving 238 bytes for local variables. Maximum is 256 bytes.
<Mitch>Next line is the beginning of the upload command line 
C:\Users\Mitch Bogart\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1/python3 -u C:\Users\Mitch Bogart\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.5.10/tools/prog.py -t uart -u COM5 -b 460800 -d attiny402 --fuses 2:0x02 6:0x04 8:0x00 -fC:\Users\MITCHB~1\AppData\Local\Temp\arduino_build_750162/Ctrlr2_2_3wire_wLight_ATtiny402_ex3_wStaggerLoad-option.ino.hex -a write -v
<Mitch>Previous line was the end of the upload command line 
SerialUPDI
UPDI programming for Arduino using a serial adapter
Based on pymcuprog, with significant modifications
By Quentin Bolsee and Spence Konde
Version 1.2.3 - Jan 2022
Using serial port COM5 at 460800 baud.
Target: attiny402
Set fuses: ['2:0x02', '6:0x04', '8:0x00']
Action: write
File: C:\Users\MITCHB~1\AppData\Local\Temp\arduino_build_750162/Ctrlr2_2_3wire_wLight_ATtiny402_ex3_wStaggerLoad-option.ino.hex
pymcuprog.programmer - INFO - Setting up programming session for 'attiny402'
pymcuprog.deviceinfo.deviceinfo - INFO - Looking for device attiny402
pymcuprog.serialupdi.physical - INFO - Opening port 'COM5' at '115200' baud
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - INFO - UPDI init OK
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - Setting UPDI clock to 16 MHz
pymcuprog.serialupdi.link - INFO - STCS 01 to 0x09
pymcuprog.serialupdi.physical - INFO - Switching to '460800' baud
pymcuprog.serialupdi.application - INFO - SIB: 'tinyAVR P:0D:0-3M2 (01.59B14.0)'
pymcuprog.serialupdi.application - INFO - Device family ID: 'tinyAVR'
pymcuprog.serialupdi.application - INFO - NVM interface: 'P:0'
pymcuprog.serialupdi.application - INFO - Debug interface: 'D:0'
pymcuprog.serialupdi.application - INFO - PDI oscillator: '3M2'
pymcuprog.serialupdi.application - INFO - Extra info: '(01.59B14.0)'
pymcuprog.serialupdi.application - INFO - Using 16-bit UPDI
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.application - INFO - PDI revision = 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.application - INFO - Entering NVM programming mode
pymcuprog.serialupdi.link - INFO - LDCS from 0x07
pymcuprog.serialupdi.application - INFO - Apply reset
pymcuprog.serialupdi.link - INFO - STCS 59 to 0x08
pymcuprog.serialupdi.application - INFO - Release reset
pymcuprog.serialupdi.link - INFO - STCS 00 to 0x08
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.nvm - INFO - No specific initializer for this provider
Pinging device...
pymcuprog.programmer - INFO - Reading device ID...
pymcuprog.serialupdi.application - INFO - SIB: 'tinyAVR P:0D:0-3M2 (01.59B14.0)'
pymcuprog.serialupdi.application - INFO - Device family ID: 'tinyAVR'
pymcuprog.serialupdi.application - INFO - NVM interface: 'P:0'
pymcuprog.serialupdi.application - INFO - Debug interface: 'D:0'
pymcuprog.serialupdi.application - INFO - PDI oscillator: '3M2'
pymcuprog.serialupdi.application - INFO - Extra info: '(01.59B14.0)'
pymcuprog.serialupdi.application - INFO - Using 16-bit UPDI
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.application - INFO - PDI revision = 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.application - INFO - Device ID from pyupdi = '1E9227' rev 'B'
pymcuprog.nvm - INFO - Device ID: '1E9227'
pymcuprog.nvm - INFO - Device revision: 'B'
pymcuprog.nvm - INFO - Device serial number: 'b'515248304d2044053658''
Ping response: 1E9227
Setting fuse 0x2=0x2
Writing literal values...
pymcuprog.programmer - INFO - Write...
pymcuprog.programmer - INFO - Writing 1 bytes of data to fuses...
pymcuprog.programmer - INFO - Write complete.
Verifying literal values...
pymcuprog.programmer - INFO - Reading 1 bytes from fuses...
pymcuprog.programmer - INFO - Verifying...
Action took 0.03s
Setting fuse 0x6=0x4
Writing literal values...
pymcuprog.programmer - INFO - Write...
pymcuprog.programmer - INFO - Writing 1 bytes of data to fuses...
pymcuprog.programmer - INFO - Write complete.
Verifying literal values...
pymcuprog.programmer - INFO - Reading 1 bytes from fuses...
pymcuprog.programmer - INFO - Verifying...
Action took 0.03s
Setting fuse 0x8=0x0
Writing literal values...
pymcuprog.programmer - INFO - Write...
pymcuprog.programmer - INFO - Writing 1 bytes of data to fuses...
pymcuprog.programmer - INFO - Write complete.
Verifying literal values...
pymcuprog.programmer - INFO - Reading 1 bytes from fuses...
pymcuprog.programmer - INFO - Verifying...
Action took 0.03s
Finished writing fuses.
Chip/Bulk erase,
Memory type eeprom is conditionally erased (depending upon EESAVE fuse setting)
Memory type flash is always erased
Memory type lockbits is always erased
...
pymcuprog.programmer - INFO - Erase...
pymcuprog.serialupdi.nvm - INFO - Chip erase using NVM CTRL
Erased.
Action took 0.01s
Writing from hex file...
Writing flash...
pymcuprog.programmer - INFO - Write...
pymcuprog.programmer - INFO - Writing 1220 bytes of data to flash...
[==================================================] 20/20
pymcuprog.programmer - INFO - Write complete.
Action took 0.24s
Verifying...
pymcuprog.programmer - INFO - Reading 1220 bytes from flash...
[==================================================] 3/3
pymcuprog.programmer - INFO - Verifying...
Verify successful. Data in flash matches data in specified hex-file
Action took 0.07s
pymcuprog.serialupdi.application - INFO - Leaving NVM programming mode
pymcuprog.serialupdi.application - INFO - Apply reset
pymcuprog.serialupdi.link - INFO - STCS 59 to 0x08
pymcuprog.serialupdi.application - INFO - Release reset
pymcuprog.serialupdi.link - INFO - STCS 00 to 0x08
pymcuprog.serialupdi.link - INFO - STCS 0C to 0x03
pymcuprog.serialupdi.physical - INFO - Closing port 'COM5'
